src/main/scala/com/softwaremill/realworld/articles/ArticlesEndpoints.scala

```diff
@@ -1,34 +1,64 @@
 package com.softwaremill.realworld.articles
 
-import com.softwaremill.realworld.profiles.ProfilesRepository
-import sttp.tapir.*
+import com.softwaremill.realworld.db.{Db, DbConfig}
+import com.softwaremill.realworld.utils.Exceptions
+import io.getquill.SnakeCase
+import sttp.tapir.PublicEndpoint
 import sttp.tapir.generic.auto.*
 import sttp.tapir.json.zio.jsonBody
 import sttp.tapir.server.ServerEndpoint.Full
-import sttp.tapir.ztapir.ZServerEndpoint
-import zio.{Cause, Exit, ZIO, ZLayer}
+import sttp.tapir.ztapir.*
 import zio.json.{DeriveJsonDecoder, DeriveJsonEncoder}
+import zio.{Cause, Exit, ZIO, ZLayer}
 
-object ArticlesEndpoints:
+import javax.sql.DataSource
 
-  given articleAuthorEncoder: zio.json.JsonEncoder[ArticleAuthor] = DeriveJsonEncoder.gen[ArticleAuthor]
-  given articleAuthorDecoder: zio.json.JsonDecoder[ArticleAuthor] = DeriveJsonDecoder.gen[ArticleAuthor]
-  given articleEncoder: zio.json.JsonEncoder[Article] = DeriveJsonEncoder.gen[Article]
-  given articleDecoder: zio.json.JsonDecoder[Article] = DeriveJsonDecoder.gen[Article]
+class ArticlesEndpoints(articlesService: ArticlesService):
+
+  import ArticlesEndpoints.given
 
   // TODO add filtering
   // TODO add pagination
-  private val list: PublicEndpoint[Unit, String, List[Article], Any] = endpoint.get
+  val list: ZServerEndpoint[DataSource, Any] = endpoint.get
     .in("api" / "articles")
     .out(jsonBody[List[Article]])
+    // TODO return proper status codes
+    // TODO format errors as json
+    .errorOut(stringBody)
+    .zServerLogic { _ =>
+      articlesService
+        .list()
+        .logError
+        .mapError(_ => "Internal error occurred.")
+    }
+
+  val get: ZServerEndpoint[DataSource, Any] = endpoint.get
+    .in("api" / "articles" / path[String]("slug"))
+    .out(jsonBody[Article])
+    // TODO return proper status codes
+    // TODO format errors as json
     .errorOut(stringBody)
+    .zServerLogic { slug =>
+      articlesService
+        .find(slug)
+        .logError
+        .mapError {
+          case Exceptions.NotFound(msg) => msg
```

Comment:
maybe this could be handled globally by customising the exception interceptor?

Reply:
I'll look into it.

Reply:
I've used the exception interceptor, but it can handle only zio defects. So I've promoted `Exceptions.NotFound` to one.

Reply:
I've created the exception handler `GlobalLoggingDefectHandler` and promoted `Exceptions.NotFound` to zio defect.